1-Draft/RFCNNNN-Notification-On-Version-Update.md

```diff
@@ -0,0 +1,242 @@
+            //    or rename the existing `_update_<old-version>` to `_update_<new-version>`.
+            // ... more ...
+
+            // Finally, create the `todayDoneFilePath` file as an indicator that a successful update check has finished today.
+            new FileStream(todayDoneFilePath, FileMode.CreateNew, FileAccess.Write, FileShare.None).Close();
+        }
+    }
+    catch (Exception e)
+    {
+        // An update check is in progress from another `pwsh` process. So it's OK to just return.
+    }
+}
+```
+
+With the file lock, only one process can get in the guarded `using` block at a given time.
+So only one process will be creating the file `_update_<version>_<publish-date>`, or renaming an old such file to reflect the new version.
+Yes, other processes could be looking at the old file name (when a `pwsh` session tries to print a notification),
+or working with an outdated `FileInfo` object (another update check tries to do a pre-validation).
+But it's fine for that to happen:
+
+- In the former case, that particular `pwsh` session will show a notification about an outdated version,
+  but the next `pwsh` session will show the right notification.
+- In the latter case, the other update check will continue and find the `.done` file already exists for today.
+
+It's possible that a `pwsh` session terminates while the update check task is still running,
+in the middle of the `using` block for example.
+Creating the `.done` file is the very last step in the `using` block.
+So if the session ends before the `.done` file is created,
+another update check will happen when the next `pwsh` session starts and finish the work.
+
+#### How to do the update check
+
+This is comparatively the easy part.
+
+- Determine if we need to check pre-releases.
+- Send HTTP query request and parse the response.
+  Some optimization work is needed in this step (see below).
+  It would be much better if we can have the latest release/pre-release information stored in a well-known URL,
+  to make the query easier and take less time.
+  - GitHub API doesn't support querying for the latest pre-release,
+    so we need to hit the 'get-all-releases' API `https://api.github.com/repos/PowerShell/PowerShell/releases`.
+    By default that will return 30 records per page and result in very expensive payload.
+    As an optimization, we should add `?per_page=4` to make it only return the most recent 4 records.
+    Most likely, they will include the latest release or pre-release.
+  - The JSON payload for 4 release records is still a lot,
+    and thus the deserialization is expensive, taking about 650 ms on my dev machine.
+    We only care about the `tag_name` and `published_at` attributes,
+    so it would be desirable to optimize the deserialization to skip the unneeded.
+- If there is a new update, create the file `_update_<version>_<publish-date>` if one doesn't exists yet;
```

Comment:
As I mentioned above I think we could pack all information in name of one file.

Reply:
I explained this in the reply to the above comment.